templates/etc/pam.d/auth.epp

```diff
@@ -0,0 +1,198 @@
+  Optional[Integer[0]]           $cracklib_minlen,
+  Integer[0]                     $cracklib_retry,
+  Integer[0]                     $deny,
+  Boolean                        $faillock,
+  Boolean                        $display_account_lock,
+  Integer[0]                     $fail_interval,
+  Integer[0]                     $remember,
+  Integer[0]                     $remember_retry,
+  Boolean                        $remember_for_root,
+  Boolean                        $even_deny_root,
+  Integer[0]                     $root_unlock_time,
+  Pam::HashAlgorithm             $hash_algorithm,
+  Integer[0]                     $rounds,
+  Integer[0]                     $uid,
+  Pam::AccountUnlockTime         $unlock_time,
+  Boolean                        $preserve_ac,
+  Boolean                        $use_netgroups,
+  Boolean                        $use_openshift,
+  Boolean                        $sssd,
+  Array[String[0]]               $tty_audit_users,
+  String[0]                      $separator,
+  Boolean                        $enable_separator,
+  Boolean                        $oath,
+  Integer[0]                     $oath_window
+|-%>
+#%PAM-1.0
+# This file managed by Puppet
+# User changes will be lost!
+<%
+if $even_deny_root {
+  $_even_deny_root = " even_deny_root root_unlock_time=${root_unlock_time}"
+}
+else {
+  $_even_deny_root = ''
+}
+-%>
+auth     optional      pam_faildelay.so
+auth     required      pam_env.so
+<%
+if $faillock {
+  if $display_account_lock {
+    $_display_account_lock = ''
+  }
+  else {
+    $_display_account_lock = ' silent'
+  }
+-%>
+auth     required      pam_faillock.so preauth<%= $_display_account_lock %> deny=<%= $deny %> audit unlock_time=<%= $unlock_time %> fail_interval=<%= $fail_interval %><%= $_even_deny_root %>
+<% } -%>
+<% if $name == 'fingerprint' { -%>
+auth     sufficient    pam_fprintd.so
+<% } -%>
+<% if $name == 'smartcard' { -%>
+auth     [success=done ignore=ignore default=die] pam_pkcs11.so wait_for_card card_only
+<% } -%>
+<% if $oath and ($name == 'system') { -%>
+auth     [success=3 default=ignore] pam_listfile.so item=group sense=allow file=/etc/liboath/exclude_groups.oath
+auth     [success=2 default=ignore] pam_listfile.so item=user sense=allow file=/etc/liboath/exclude_users.oath
+auth     [success=1 default=bad]    pam_oath.so usersfile=/etc/liboath/users.oath window=<%= $oath_window %>
+auth     requisite     pam_deny.so
+<% } -%>
+<% if $name in ['system', 'password'] { -%>
+<%   if $sssd { -%>
+auth     sufficient    pam_sss.so forward_pass
+auth     sufficient    pam_unix.so try_first_pass
+<%   } else { -%>
+auth     sufficient    pam_unix.so try_first_pass
+<%   } -%>
+<%   if $faillock { -%>
+auth     [default=die] pam_faillock.so authfail deny=<%= $deny %> audit unlock_time=<%= $unlock_time %><%= $_even_deny_root %>
+auth     sufficient    pam_faillock.so authsucc deny=<%= $deny %> audit unlock_time=<%= $unlock_time %><%= $_even_deny_root %>
```

Comment:
If pass_sss.so or pam_unix.so above are 'sufficient', you will never get to this line.  So, I don't see how the faillock count will be cleared.  Am I missing something?

Reply:
Honestly, I don't know how it was getting cleared anyway, perhaps the `account` line.

Anyway, updating for proper ordering but going to need some hands-on testing somewhere.